feat: add C implementation for `stats/base/dists/geometric/pmf`

0PrashantYadav0 opened the pull request:
Resolves #3639

## Description

> What is the purpose of this pull request?

This pull request:

-   adds C implementation for @stdlib/stats/base/dists/geometric/pmf along with relevant tests, benchmarks and examples.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #3639

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

0PrashantYadav0 committed:
feat: add C implementation for geometric pmf

0PrashantYadav0 committed:
fix: change example in readme

0PrashantYadav0 committed:
fix: changes in benchmark and test

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/geometric/pmf/benchmark/benchmark.js:
```diff
@@ -32,16 +33,23 @@ var pmf = require( './../lib' );
+	p = new Float64Array( len );
+	for ( i = 0; i < len; i++ ) {
+		x[ i ] = round( randu()*40.0 );
+		p[ i ] = ( randu()*1.0 ) + EPS;
```
Let's not add `EPS` here because `p` should be a number in `[0,1]`.

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/geometric/pmf/README.md:
```diff
@@ -133,6 +133,104 @@ for ( i = 0; i < 10; i++ ) {
+
+#### stdlib_base_dists_geometric_pmf( x, p )
+
+Evaluates evaluating the [probability mass function][pmf] (PMF) of a [geometric][geometric-distribution] distribution with success probability `0 <= p <= 1`.
```
```suggestion
Evaluates the [probability mass function][pmf] (PMF) of a [geometric][geometric-distribution] distribution with success probability `0 <= p <= 1`.
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/geometric/pmf/README.md:
```diff
@@ -133,6 +133,104 @@ for ( i = 0; i < 10; i++ ) {
+
+    for ( i = 0; i < 25; i++ ) {
+        x = stdlib_base_round( random_uniform( 0, 40 ) );
+        p = random_uniform( 0.0, 1.0 ) + STDLIB_CONSTANT_FLOAT64_EPS;
```
Let's not add `EPS` here because `p` should be a number in `[0,1]`.

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/geometric/pmf/benchmark/benchmark.native.js:
```diff
@@ -0,0 +1,72 @@
+	p = new Float64Array( len );
+	for ( i = 0; i < len; i++ ) {
+		x[ i ] = round( randu()*40.0 );
+		p[ i ] = ( randu()*1.0 ) + EPS;
```
Let's not add `EPS` here because `p` should be a number in `[0,1]`.

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/geometric/pmf/benchmark/c/benchmark.c:
```diff
@@ -0,0 +1,142 @@
+
+	for ( i = 0; i < 100; i++ ) {
+		x[ i ] = stdlib_base_round( random_uniform( 0.0, 40.0 ) );
+		p[ i ] = random_uniform( 0.0, 1.0 ) + STDLIB_CONSTANT_FLOAT64_EPS;
```
Let's not add `EPS` here because `p` should be a number in `[0,1]`.

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/geometric/pmf/include/stdlib/stats/base/dists/geometric/pmf.h:
```diff
@@ -0,0 +1,38 @@
+#endif
+
+/**
+* Evaluates the probability mass function (PMF) for the geometric distribution with probability parameter p.
```
```suggestion
* Evaluates the probability mass function (PMF) for a geometric distribution with success probability `p` at a value `x`.
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/geometric/pmf/manifest.json:
```diff
@@ -0,0 +1,86 @@
+{
+    "options": {
```
Two-space indentation needed here.

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/geometric/pmf/src/main.c:
```diff
@@ -0,0 +1,70 @@
+* // returns NaN
+*/
+double stdlib_base_dists_geometric_pmf( const double x, const double p ) {
+    if (
```
Should be tab-indented.

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/geometric/pmf/src/main.c:
```diff
@@ -0,0 +1,70 @@
+/**
+* Evaluates the probability mass function (PMF) for a geometric distribution with success probability `p` at a value `x`.
+*
+* @param {number} x - input value
```
Shouldn't have JSDoc comments in C files. No type annotations and descriptions should be vertically aligned.

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/geometric/pmf/src/main.c:
```diff
@@ -0,0 +1,70 @@
+#include "stdlib/math/base/assert/is_nonnegative_integer.h"
+#include "stdlib/math/base/assert/is_nan.h"
+#include "stdlib/math/base/special/pow.h"
+#include <math.h>
```
Not used:

```suggestion
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/geometric/pmf/src/main.c:
```diff
@@ -0,0 +1,70 @@
+* @param {Probability} p - success probability
+* @returns {Probability} evaluated PMF
+*
+* @example
+* var y = pmf( 4.0, 0.3 );
+* // returns ~0.072
+*
+* @example
+* var y = pmf( 2.0, 0.7 );
+* // returns ~0.063
+*
+* @example
+* var y = pmf( -1.0, 0.5 );
+* // returns 0.0
+*
+* @example
+* var y = pmf( 0.0, NaN );
+* // returns NaN
+*
+* @example
+* var y = pmf( NaN, 0.5 );
+* // returns NaN
+*
+* @example
+* // Invalid success probability:
+* var y = pmf( 2.0, 1.5 );
+* // returns NaN
```
Let's just keep one example; also needs to be updated for C (no `var`):

```suggestion
* @example
* double y = pmf( 4.0, 0.3 );
* // returns ~0.072
```

Planeshifter reviewed (changes requested):
Did an initial review and left comments that should be addressed. Thanks!

0PrashantYadav0 committed:
fix: remove eps and changes intendation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: passed
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: passed
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: passed
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

0PrashantYadav0 committed:
fix: changes in manifest

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

0PrashantYadav0 commented:
@Planeshifter I made changes in this pr as well as in all the pr related to geometric distribution. Please look into all of these.

Planeshifter commented:
/stdlib lint-autofix

stdlib-bot committed:
fix: resolve lint errors

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/geometric/pmf/src/main.c:
```diff
@@ -0,0 +1,48 @@
+	) {
+		return 0.0/0.0; // NaN
+	}
+	if( stdlib_base_is_nonnegative_integer( x ) ){
```
```suggestion
	if ( stdlib_base_is_nonnegative_integer( x ) ) {
```

Planeshifter reviewed (approved):
Thanks! Let's land this PR shortly.

Planeshifter committed:
Update lib/node_modules/@stdlib/stats/base/dists/geometric/pmf/src/main.c

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>
